components/homme/src/theta-l_kokkos/cxx/HyperviscosityFunctorImpl.hpp

```diff
@@ -278,25 +351,89 @@ class HyperviscosityFunctorImpl
                            [&](const int ilev) {
-        utens(ilev)   *= m_data.dt*rspheremp;
-        vtens(ilev)   *= m_data.dt*rspheremp;
-        ttens(ilev)   *= m_data.dt*rspheremp;
-        dptens(ilev)  *= m_data.dt*rspheremp;
+
+//introduce a var for dt/hvs			   
+        utens(ilev)   *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
+        vtens(ilev)   *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
+        ttens(ilev)   *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
+        dptens(ilev)  *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
 
         u(ilev)      += utens(ilev);
         v(ilev)      += vtens(ilev);
         vtheta(ilev) += ttens(ilev);
         dp(ilev)     += dptens(ilev);
         if (m_process_nh_vars) {
-          wtens(ilev)   *= m_data.dt*rspheremp;
-          phitens(ilev) *= m_data.dt*rspheremp;
+          wtens(ilev)   *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
+          phitens(ilev) *= (m_data.dt/m_data.hypervis_subcycle)*rspheremp;
 
           w(ilev)      += wtens(ilev);
           phi_i(ilev)  += phitens(ilev);
         }
       });
     });
-  }
+  }  //tagupdatestates
+
+
+
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const TagUpdateStates2&, const TeamMember& team) const {
+    KernelVariables kv(team, m_tu);
+
+    using MidColumn = decltype(Homme::subview(m_buffers.wtens,0,0,0));
+    using IntColumn = decltype(Homme::subview(m_state.m_w_i,0,0,0,0));
+
+//define another dt
+//    m_data.dt_top = 
+
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team,NP*NP),
+                         [&](const int idx) {
+      const int igp = idx / NP;
+      const int jgp = idx % NP;
+
+      // Add Xtens quantities back to the states, except for vtheta
+      auto u = Homme::subview(m_state.m_v,kv.ie,m_data.np1,0,igp,jgp);
+      auto v = Homme::subview(m_state.m_v,kv.ie,m_data.np1,1,igp,jgp);
+      auto vtheta = Homme::subview(m_state.m_vtheta_dp,kv.ie,m_data.np1,igp,jgp);
+      auto dp     = Homme::subview(m_state.m_dp3d,kv.ie,m_data.np1,igp,jgp);
+
+      auto utens   = Homme::subview(m_buffers.vtens,kv.ie,0,igp,jgp);
+      auto vtens   = Homme::subview(m_buffers.vtens,kv.ie,1,igp,jgp);
+      auto ttens   = Homme::subview(m_buffers.ttens,kv.ie,igp,jgp);
+      auto dptens  = Homme::subview(m_buffers.dptens,kv.ie,igp,jgp);
+      const auto& rspheremp = m_geometry.m_rspheremp(kv.ie,igp,jgp);
+
+      MidColumn wtens, phitens;
+      IntColumn w, phi_i;
+
+      if (m_process_nh_vars) {
+        wtens   = Homme::subview(m_buffers.wtens,kv.ie,igp,jgp);
+        phitens = Homme::subview(m_buffers.phitens,kv.ie,igp,jgp);
+        w       = Homme::subview(m_state.m_w_i,kv.ie,m_data.np1,igp,jgp);
+        phi_i   = Homme::subview(m_state.m_phinh_i,kv.ie,m_data.np1,igp,jgp);
+      }
+
+      Kokkos::parallel_for(Kokkos::ThreadVectorRange(kv.team,NUM_LEV),
+                           [&](const int ilev) {
+        utens(ilev)   *= rspheremp;
```

Comment:
@ambrad Do you think it might be beneficial to add fmad-like operators/functions updates for our Pack/Vector classes, to allow doing
```
u(ilev) = u(ilev) + utens(ilev)*rspheremp
```
as a single fmad op, rather than have one loop for the `*= rspheremp` part, and one loop for the `+=` part? I doubt the compiler can fuse those two loops into one, and fuse + and * (either on CPU or GPU).

Reply:
I am confused, I do not see 2 loops there. I probably broke off + and * just to make it bfb, without knowing that I cannot do it in 1 line on vectors.

Reply:
Sorry, I wasn't clear. Your code is fine. Separate the * and the + is the way to do it, _given our Vector class impl_. The loops I referred to are _hidden_ inside the Vector class. When you do `a *= b`, with `a` being a Vector, the overload of `*=` includes a for loop over the Vector length. So, when we do
```
utens(ilev) *= rspheremp;
u(ilev) += utens(ilev);
```
we are doing 2 for loops. And I doubt that the compiler can fuse those loops into one, and merge the *= with the +=, into a single fmad op.

Reply:
I'm actually quite sure the compiler cannot do FMA here, since it cannot know that we are not really interesting in storing `utens*rspheremp` inside utens, so it must keep the `*=` separate from the `+=`.

TBC: this impl is what we are currently doing throughout Hommexx. I am just wondering if there is room for a micro opt here.

Reply:
I haven't looked whether `u(ilev) += rspheremp*utens(ilev)` gets compiled to an FMA. I think it's possible with the ekat pack impl but unlikely with the Hxx one. Not reason not to add an explicit FMA routine.